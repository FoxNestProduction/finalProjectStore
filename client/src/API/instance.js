import axios from 'axios';

let store;

export const injectStore = (_store) => {
  store = _store;
};

export const instance = axios.create({
  baseURL: process.env.REACT_APP_API_URL,
});

instance.interceptors.request.use((config) => {
  if (store.getState().authorization && store.getState().authorization.token) {
    // eslint-disable-next-line no-param-reassign
    config.headers.authorization = store.getState().authorization.token;
  }
  return config;
}, (error) => {
  return Promise.reject(error);
});
