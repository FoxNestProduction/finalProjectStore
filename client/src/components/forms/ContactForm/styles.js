export const formWrapper = {
  position: 'relative',
  display: 'flex',
  flexDirection: 'column',
  flexGrow: 1,
  m: { mobile: '40px auto 0', lgTablet: '0' },
  width: {
    mobile: '100%',
    tablet: '90%',
    lgTablet: '100%',
  },
  maxWidth: {
    tablet: '380px',
    lgTablet: '100%',
  },
  minWidth: {
    lgTablet: '260px',
  },
  bgcolor: 'transparent',
};

export const mainTitle = {
  display: {
    mobile: 'none',
    lgTablet: 'block',
  },
  mb: {
    lgTablet: '34px',
    desktop: '27px',
  },
  textAlign: 'center',
  fontWeight: 'fontWeightMedium',
  fontSize: {
    mobile: '26px',
    tablet: '28px',
    lgTablet: '36px',
    desktop: '42px',
  },
  color: 'text.primary',
};

export const inputsWrapper = {
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'stretch',
  gap: {
    mobile: '20px;',
    tablet: '17px',
    desktop: '24px',
  },
  mb: {
    mobile: '10px',
    tablet: '9px',
    desktop: '24px',
  },
  width: '100%',
};

export const sendBtn = {
  width: '100%',
  mt: '12px',
  height: {
    mobile: '60px',
    tablet: '53px',
    desktop: '60px',
  },
  color: 'text.primaryLight',
  transition: 'background-color 0.3s ease, box-shadow 0.3s ease, color 0.3s ease',
  '&:hover': {
    backgroundColor: 'primary.hover',
    boxShadow: '0px 4px 6px rgba(0, 0, 0, 0.3)',
  },
  '&:active': {
    boxShadow: '0px -1px 4px rgba(0, 0, 0, 0.5)',
    transform: 'translateY(1px)',
    backgroundColor: 'common.white',
    color: '#1C186C',
    border: '1px solid primary.main',
    borderColor: 'primary.main',
  },
  fontSize: {
    mobile: '14px',
    desktop: '20px',
  },
  fontWeight: {
    mobile: 'fontWeightSemiBold',
    desktop: 'fontWeightRegular',
  },
};

export const purpleArrow = {
  display: {
    mobile: 'none',
    lgTablet: 'block',
  },
  position: 'absolute',
  bottom: '-20%',
  left: '-22%',
  transform: 'rotate(-78deg)',
};
