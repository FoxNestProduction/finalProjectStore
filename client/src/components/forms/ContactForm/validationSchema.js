import { object, string } from 'yup';

const validationSchema = object({
  name: string()
    .required('This field is required')
    .min(2, 'Name is too short'),
  email: string()
    .required('Email is required')
    .email('Invalid email format'),
  message: string()
    .required('This field is required')
    .min(10, 'Message is too short'),
});

export default validationSchema;
