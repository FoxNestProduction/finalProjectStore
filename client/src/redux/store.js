import { configureStore, getDefaultMiddleware, combineReducers } from '@reduxjs/toolkit';
// todo: правила для redux-persist
// eslint-disable-next-line import/no-extraneous-dependencies
import storage from 'redux-persist/lib/storage';
// eslint-disable-next-line import/no-extraneous-dependencies
import { persistReducer } from 'redux-persist';
import modalSlice from './slices/modalSlice';
import productsSlice from './slices/productsSlice';
import authorizationSlice from './slices/authorizationSlice';
import userSlice from './slices/userSlice';
import partnersSlice from './slices/partnersSlice';
import searchSlice from './slices/searchSlice';
import errorSlice from './slices/errorSlice';
import reviewsSlice from './slices/reviewsSlice';
import favouriteSlice from './slices/favouriteSlice';

const authPersistConfig = {
  key: 'authorization',
  version: 1,
  storage,
};

const userPersistConfig = {
  key: 'user',
  version: 1,
  storage,
};

const searchPersistConfig = {
  key: 'search',
  version: 1,
  storage,
};
const favouritePersistConfig = {
  key: 'favorites',
  version: 1,
  storage,
};

/*
    todo: - ЗВЕРНІТЬ УВАГУ!
      якщо вам потрібно щоб ваш стейт не зникав після перезавантаження
      то юзайте підхід як authorization і user,
      якщо ж ні то варіант написання як для modal всередині reducer
      Якщо що пишіть @Ihor_Kacher
*/
const reducer = combineReducers({
  error: errorSlice,
  modal: modalSlice,
  partners: partnersSlice,
  authorization: persistReducer(authPersistConfig, authorizationSlice),
  user: persistReducer(userPersistConfig, userSlice),
  products: productsSlice,
  search: persistReducer(searchPersistConfig, searchSlice),
  reviews: reviewsSlice,
  favourites: persistReducer(favouritePersistConfig, favouriteSlice),
});

const middleware = getDefaultMiddleware({
  serializableCheck: false,
});

const store = configureStore({
  reducer,
  middleware,
});

export default store;
