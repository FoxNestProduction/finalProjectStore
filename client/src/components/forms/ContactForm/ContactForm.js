import React, { memo } from 'react';
import { Formik, Form } from 'formik';
import Typography from '@mui/material/Typography';
import Box from '@mui/material/Box';
import Button from '@mui/material/Button';
import validationSchema from './validationSchema';
import {
  mainTitle,
  inputsWrapper,
  sendBtn,
  formWrapper,
  purpleArrow,
} from './styles';
import Input from '../../inputs/Input/Input';
import Textarea from '../../inputs/Textarea/Textarea';
import OneLoopArrowSvg from '../../../assets/svgComponents/OneLoopArrowSvg';

const ContactForm = () => {
  const initialValues = {
    name: '',
    email: '',
    message: '',
  };

  const handleSubmit = (values, actions) => {
    console.log(values);
    // actions.resetForm();
  };

  return (
    <Box sx={formWrapper}>
      <Typography variant="h2" component="h1" sx={mainTitle}>
        Customer Support
      </Typography>
      <Formik
        initialValues={initialValues}
        onSubmit={handleSubmit}
        validationSchema={validationSchema}
      >
        {({ isValid }) => (
          <Form>
            <Box sx={inputsWrapper}>
              <Input
                type="text"
                name="name"
                id="contactName"
                placeholder="Enter your name"
                label="Name"
                bgColor="#FFF"
              />
              <Input
                type="email"
                name="email"
                id="contactEmail"
                placeholder="Enter your e-mail"
                label="E-mail"
                bgColor="#FFF"
              />
              <Textarea name="message" placeholder="Enter the problem or query..." />
            </Box>
            <Box sx={{ textAlign: 'center' }}>
              <Button
                disableRipple
                variant="contained"
                sx={sendBtn}
                type="submit"
                disabled={!isValid}
              >
                Send Now
              </Button>
            </Box>
          </Form>
        )}
      </Formik>
      <Box sx={purpleArrow}>
        <OneLoopArrowSvg color="#6C5FBC" />
      </Box>
    </Box>
  );
};

export default memo(ContactForm);
