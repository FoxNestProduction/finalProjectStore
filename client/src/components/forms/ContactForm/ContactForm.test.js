import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../../../redux/store';
import ContactForm from './ContactForm';
import { instance } from '../../../API/instance';

jest.mock('../../../API/instance', () => ({
  instance: {
    post: jest.fn(),
  },
}));

describe('Contact form component', () => {
  test('contact form snapshot', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <ContactForm />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('try to submit empty form', async () => {
    render(
      <Provider store={store}>
        <ContactForm />
      </Provider>,
    );

    const submitButton = screen.getByText('Send Now');
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(screen.getByText('Email is required')).toBeInTheDocument();
    });
    await waitFor(() => {
      expect(submitButton).toBeDisabled();
    });
  });

  test('Submit correct form', async () => {
    render(
      <Provider store={store}>
        <ContactForm />
      </Provider>,
    );

    const nameInput = screen.getByPlaceholderText('Enter your name');
    const emailInput = screen.getByPlaceholderText('Enter your e-mail');
    const textarea = screen.getByPlaceholderText('Enter the problem or query...');
    const submitButton = screen.getByText('Send Now');

    fireEvent.change(nameInput, { target: { value: 'Yurii' } });
    fireEvent.change(emailInput, { target: { value: 'example@example.com' } });
    fireEvent.change(textarea, { target: { value: 'Some message' } });

    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(instance.post).toHaveBeenCalledTimes(1);
    });
    await waitFor(() => {
      const value = {
        name: 'Yurii',
        email: 'example@example.com',
        message: 'Some message',
      };
      expect(instance.post).toHaveBeenCalledWith('/support', value);
    });
  });
});
