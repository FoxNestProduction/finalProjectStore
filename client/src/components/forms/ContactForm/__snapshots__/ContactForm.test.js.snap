// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact form component contact form snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-hn2bkr"
  >
    <h1
      class="MuiTypography-root MuiTypography-h2 css-19nzqo3-MuiTypography-root"
    >
      Customer Support
    </h1>
    <form
      action="#"
    >
      <div
        class="MuiBox-root css-1snh8b2"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="contactName"
          >
            Name
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1dnlmzz-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-describedby="helper-text"
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="contactName"
              name="name"
              placeholder="Enter your name"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Name
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
            id="helper-text"
          />
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="contactEmail"
          >
            E-mail
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1dnlmzz-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-describedby="helper-text"
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="contactEmail"
              name="email"
              placeholder="Enter your e-mail"
              type="email"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  E-mail
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
            id="helper-text"
          />
        </div>
        <div
          class="root"
        >
          <textarea
            class="textarea"
            name="message"
            placeholder="Enter the problem or query..."
            rows="6"
          />
        </div>
      </div>
      <div
        class="MuiBox-root css-xi606m"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-xheoip-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Send Now
        </button>
      </div>
    </form>
    <div
      class="MuiBox-root css-18pdv8v"
    >
      <svg
        fill="none"
        height="118"
        viewBox="0 0 117 118"
        width="117"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M57.9545 94.9949C47.2252 85.339 38.8069 73.1896 38.0046 58.5059C37.2806 45.2521 44.3231 35.2379 58.4214 41.1439C65.0953 43.9396 71.4587 51.6393 70.3128 59.0609C69.0362 67.3292 58.5027 63.3819 54.5536 59.1501C45.3299 49.2665 43.4373 32.8851 52.6569 22.5675C63.5751 10.3491 74.9329 24.3724 74.8255 24.4077"
          stroke="#6C5FBC"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-miterlimit="1.5"
          stroke-width="2.09329"
        />
        <path
          d="M42.5719 95.8242C43.8782 95.0961 45.3767 95.0149 46.8626 94.8559C51.7416 94.3332 57.2693 95.3597 61.4869 98.4709C58.0966 93.8961 55.1476 86.9063 57.1738 81.6141"
          stroke="#6C5FBC"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-miterlimit="1.5"
          stroke-width="2.09329"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
